File: ext/examples/bruner.rs
//! This script converts between our basis and Bruner's basis. At the moment, most inputs are
//! hardcoded, and this only works for the sphere.
//!
//! The script performs the following procedure:
//!
//! 1. Compute our own resolution with the Milnor basis
//! 2. Create Bruner's resolution as a
//!    [`FiniteChainComplex`](ext::chain_complex::FiniteChainComplex) object
//! 3. Use a [`ResolutionHomomorphism`](ext::resolution_homomorphism::ResolutionHomomorphism) to
//!    lift the identity to a chain map from Bruner's resolution
//!    to our resolution. We should do it in this direction because we have stored the
//!    quasi-inverses for our resolution, but not Bruner's.
//! 4. Read off the transformation matrix we need
//!
//! The main extra work to put in is step (2), where we have to parse Bruner's differentials and
//! interpret it as a chain complex. Bruner's resolution can be found at
//! <https://archive.sigma2.no/pages/public/datasetDetail.jsf?id=10.11582/2022.00015>
//! while the descirption of his save file is at <https://arxiv.org/abs/2109.13117>.

use std::{
    fs::File,
    io,
    path::{Path, PathBuf},
    str::FromStr,
    sync::Arc,
};

use algebra::{
    milnor_algebra::MilnorBasisElement,
    module::{homomorphism::FreeModuleHomomorphism as FMH, FreeModule as FM, Module},
    Algebra, MilnorAlgebra,
};
use anyhow::{Context, Error, Result};
use ext::{
    chain_complex::{ChainComplex, FiniteChainComplex as FCC},
    resolution_homomorphism::ResolutionHomomorphism,
};
use fp::{matrix::Matrix, prime::TWO, vector::FpVector};
use sseq::coordinates::{Bidegree, BidegreeGenerator};

#[cfg(feature = "nassau")]
type FreeModule = FM<MilnorAlgebra>;
#[cfg(not(feature = "nassau"))]
type FreeModule = FM<algebra::SteenrodAlgebra>;

type FreeModuleHomomorphism = FMH<FreeModule>;
type FiniteChainComplex = FCC<FreeModule, FreeModuleHomomorphism>;

/// Read the first non-empty line of `data` into `buf`. Returns whether a line is read
fn read_line(data: &mut impl io::BufRead, buf: &mut String) -> Result<bool> {
    buf.clear();
    while buf.is_empty() {
        let num_bytes = data.read_line(buf)?;
        if num_bytes == 0 {
            return Ok(false);
        }
        // Remove newline character
        buf.pop();
    }
    Ok(true)
}

/// Viewing `s` as a whitespace-delimited array, take the first item and parse it into T.
fn entry<T>(x: &str) -> Result<(&str, T)>
where
    T: FromStr,
    Error: From<<T as FromStr>::Err>,
{
    let x = x.trim();
    match x.find(' ') {
        Some(k) => Ok((&x[k..], x[..k].parse()?)),
        None => Ok(("", x.parse()?)),
    }
}

/// Read an algebra element, where input contains
/// ```text
/// $op_deg _ $op
/// ```
/// This returns an iterator of indices of the operators whose sum is the element sought.
fn get_algebra_element<'a>(
    a: &'a MilnorAlgebra,
    input: &'a str,
) -> Result<impl Iterator<Item = usize> + 'a> {
    let (input, t) = entry(input)?;
    let (input, _) = entry::<u32>(input)?;

    let input = input.trim();
    assert_eq!(&input[0..1], "i");

    // Remove the i
    let input = &input[1..];
    // Remove the trailing ).
    let input = &input[..input.len() - 2];

    Ok(input.split(')').map(move |entry| {
        let entry = &entry[1..];
        let elt = MilnorBasisElement {
            q_part: 0,
            p_part: entry.split(',').map(|x| x.parse().unwrap()).collect(),
            degree: t,
        };
        a.basis_element_to_index(&elt)
    }))
}

/// Get a block describing a generator. Returns the degree and the value of the differential.
fn get_element(
    a: &MilnorAlgebra,
    m: &FreeModule,
    input: &mut impl io::BufRead,
) -> Result<Option<(i32, FpVector)>> {
    let mut buf = String::new();
    if !read_line(input, &mut buf)? {
        return Ok(None);
    }
    let degree: i32 = buf.trim().parse()?;
    a.compute_basis(degree);
    m.compute_basis(degree);

    read_line(input, &mut buf)?;
    let num_lines: usize = buf.trim().parse()?;

    let mut result = FpVector::new(TWO, m.dimension(degree));

    for _ in 0..num_lines {
        read_line(input, &mut buf)?;
        let (rem, gen_idx) = entry::<usize>(&buf)?;
        let offset = m.internal_generator_offset(degree, gen_idx);
        for op in get_algebra_element(a, &rem[1..])? {
            result.add_basis_element(offset + op, 1);
        }
    }
    Ok(Some((degree, result)))
}

/// Create a new `FiniteChainComplex` with `num_s` many non-zero modules.
fn create_chain_complex(num_s: usize) -> FiniteChainComplex {
    #[cfg(feature = "nassau")]
    let algebra: Arc<MilnorAlgebra> = Arc::new(MilnorAlgebra::new(TWO, false));

    #[cfg(not(feature = "nassau"))]
    let algebra: Arc<algebra::SteenrodAlgebra> = Arc::new(algebra::SteenrodAlgebra::MilnorAlgebra(
        MilnorAlgebra::new(TWO, false),
    ));

    let mut modules: Vec<Arc<FreeModule>> = Vec::with_capacity(num_s);
    let mut differentials: Vec<Arc<FreeModuleHomomorphism>> = Vec::with_capacity(num_s - 1);
    for _ in 0..num_s {
        modules.push(Arc::new(FreeModule::new(
            Arc::clone(&algebra),
            String::new(),
            0,
        )));
    }
    for s in 1..num_s {
        differentials.push(Arc::new(FreeModuleHomomorphism::new(
            Arc::clone(&modules[s]),
            Arc::clone(&modules[s - 1]),
            0,
        )));
    }
    FiniteChainComplex::new(modules, differentials)
}

/// Read the Diff.$N files in `data_dir` and produce the corresponding chain complex object.
fn read_bruner_resolution(data_dir: &Path, max_n: i32) -> Result<(u32, FiniteChainComplex)> {
    let num_s: usize = data_dir.read_dir()?.count();

    let cc = create_chain_complex(num_s);
    let algebra = cc.algebra();

    let algebra: &MilnorAlgebra = algebra.as_ref().try_into()?;

    let mut buf = String::new();
    let s = num_s as u32 - 1;

    algebra.compute_basis(max_n + s as i32 + 1);
    // Handle s = 0
    {
        // TODO: actually parse file
        let m = cc.module(0);
        m.add_generators(0, 1, None);
        m.extend_by_zero(max_n + 1);
    }

    for s in 1..num_s as u32 {
        let m = cc.module(s);
        let d = cc.differential(s);

        let mut f = io::BufReader::new(
            File::open(data_dir.join(format!("hDiff.{s}")))
                .with_context(|| format!("Failed to read hDiff.{s}"))?,
        );

        read_line(&mut f, &mut buf)?;

        let mut entries: Vec<FpVector> = Vec::new();
        let mut cur_degree: i32 = 0;

        while let Some((t, gen)) = get_element(algebra, cc.module(s - 1).as_ref(), &mut f)? {
            if t == cur_degree {
                entries.push(gen);
            } else {
                m.add_generators(cur_degree, entries.len(), None);
                d.add_generators_from_rows(cur_degree, entries);

                m.extend_by_zero(t - 1);
                d.extend_by_zero(t - 1);

                entries = vec![gen];
                cur_degree = t;
            }
        }
        m.add_generators(cur_degree, entries.len(), None);
        d.add_generators_from_rows(cur_degree, entries);

        m.extend_by_zero(max_n + s as i32 + 1);
        d.extend_by_zero(max_n + s as i32);
    }

    Ok((s, cc))
}

fn main() {
    ext::utils::init_logging();

    let data_dir = Path::new(file!()).parent().unwrap().join("bruner_data");
    let max_n: i32 = query::with_default("Max n", "20", str::parse);

    // Read in Bruner's resolution
    let (max_s, cc) = read_bruner_resolution(&data_dir, max_n).unwrap();
    let max = Bidegree::n_s(max_n, max_s);
    let cc = Arc::new(cc);

    let save_dir = query::optional("Save directory", |x| {
        core::result::Result::<PathBuf, std::convert::Infallible>::Ok(PathBuf::from(x))
    });

    #[cfg(feature = "nassau")]
    assert!(
        save_dir.is_some(),
        "A save directory is required for comparison between Bruner and Nassau resolutions."
    );

    let resolution = ext::utils::construct("S_2@milnor", save_dir).unwrap();

    resolution.compute_through_stem(max);

    let resolution = Arc::new(resolution);

    // Create a ResolutionHomomorphism object
    let hom = ResolutionHomomorphism::new(String::new(), cc, resolution, Bidegree::zero());

    // We have to explicitly tell it what to do at (0, 0)
    hom.extend_step(Bidegree::zero(), Some(&Matrix::from_vec(TWO, &[vec![1]])));
    hom.extend_all();

    // Now print the results
    println!("sseq_basis | bruner_basis");
    for b in hom.target.iter_stem() {
        let matrix = hom.get_map(b.s()).hom_k(b.t());

        for (i, row) in matrix.into_iter().enumerate() {
            let gen = BidegreeGenerator::new(b, i);
            println!("x_{gen:#} = {row:?}");
        }
    }
}

File: ext_py/examples/bruner.py
#!/usr/bin/env python3
"""Convert between our basis and Bruner's basis (sphere only, mostly hardcoded).

Python port of ext/examples/bruner.rs.
"""

import os
import sys

import _query as query
from ext import Resolution, algebra, fp, sseq
from ext.algebra import MilnorAlgebra, MilnorBasisElement, SteenrodAlgebra
# NOTE: depends on FreeModule, FreeModuleHomomorphism (algebra, API_PROPOSAL §5.3/§5.4),
# FiniteChainComplex (ext top level, API_PROPOSAL §7.1), ResolutionHomomorphism
# (ext top level, API_PROPOSAL §7.3), construct (ext, API_PROPOSAL §7.7).
from ext import FiniteChainComplex, ResolutionHomomorphism
# NOTE: depends on Matrix.from_vec (fp, API_PROPOSAL §4.4) and the TWO constant
# (fp, API_PROPOSAL §4.1).
from ext.fp import FpVector, Matrix

TWO = fp.TWO


def read_line(data):
    """Read the first non-empty line of ``data``; return it without the newline.

    Returns ``None`` at end of file, mirroring the Rust ``read_line`` helper.
    """
    buf = ""
    while buf == "":
        line = data.readline()
        if line == "":
            return None
        # Remove newline character (Rust pops the trailing byte unconditionally).
        buf = line[:-1]
    return buf


def entry(x):
    """Take the first whitespace-delimited item of ``x``; return ``(rest, item)``."""
    x = x.strip()
    k = x.find(" ")
    if k == -1:
        return ("", x)
    return (x[k:], x[:k])


def get_algebra_element(a, input_str):
    """Parse an algebra element ``$op_deg _ $op``; yield indices whose sum is it."""
    input_str, t = entry(input_str)
    t = int(t)
    input_str, _ = entry(input_str)

    input_str = input_str.strip()
    assert input_str[0:1] == "i"

    # Remove the i
    input_str = input_str[1:]
    # Remove the trailing ).
    input_str = input_str[: len(input_str) - 2]

    for piece in input_str.split(")"):
        piece = piece[1:]
        elt = MilnorBasisElement(
            [int(x) for x in piece.split(",")],
            0,
            t,
        )
        yield a.basis_element_to_index(elt)


def get_element(a, m, input_data):
    """Read a generator block; return ``(degree, FpVector)`` or ``None`` at EOF."""
    buf = read_line(input_data)
    if buf is None:
        return None
    degree = int(buf.strip())
    a.compute_basis(degree)
    m.compute_basis(degree)

    buf = read_line(input_data)
    num_lines = int(buf.strip())

    result = FpVector(TWO, m.dimension(degree))

    for _ in range(num_lines):
        buf = read_line(input_data)
        rem, gen_idx = entry(buf)
        gen_idx = int(gen_idx)
        offset = m.internal_generator_offset(degree, gen_idx)
        for op in get_algebra_element(a, rem[1:]):
            result.add_basis_element(offset + op, 1)
    return (degree, result)


def create_chain_complex(num_s):
    """Create a ``FiniteChainComplex`` with ``num_s`` non-zero modules."""
    alg = SteenrodAlgebra.milnor(int(TWO), False)

    modules = []
    differentials = []
    for _ in range(num_s):
        modules.append(algebra.FreeModule(alg, "", 0))
    for s in range(1, num_s):
        differentials.append(
            algebra.FreeModuleHomomorphism(modules[s], modules[s - 1], 0)
        )
    return FiniteChainComplex(modules, differentials)


def read_bruner_resolution(data_dir, max_n):
    """Read the ``hDiff.N`` files in ``data_dir`` into a chain complex.

    Returns ``(max_s, FiniteChainComplex)``.
    """
    num_s = len(os.listdir(data_dir))

    cc = create_chain_complex(num_s)
    # Rust does `cc.algebra().as_ref().try_into()` to view the chain complex's
    # SteenrodAlgebra as a &MilnorAlgebra. No SteenrodAlgebra -> MilnorAlgebra
    # conversion is bound, so we rebuild the same Milnor algebra here.
    # NOTE: depends on SteenrodAlgebra -> MilnorAlgebra conversion (API_PROPOSAL §5.2);
    # using a fresh MilnorAlgebra(2, False) as a faithful stand-in.
    alg = MilnorAlgebra(int(TWO), False)

    s = num_s - 1

    alg.compute_basis(max_n + s + 1)
    # Handle s = 0
    # TODO: actually parse file
    m = cc.module(0)
    m.add_generators(0, 1, None)
    m.extend_by_zero(max_n + 1)

    for s in range(1, num_s):
        m = cc.module(s)
        d = cc.differential(s)

        with open(os.path.join(data_dir, f"hDiff.{s}")) as f:
            read_line(f)

            entries = []
            cur_degree = 0

            while True:
                element = get_element(alg, cc.module(s - 1), f)
                if element is None:
                    break
                t, g = element
                if t == cur_degree:
                    entries.append(g)
                else:
                    m.add_generators(cur_degree, len(entries), None)
                    d.add_generators_from_rows(cur_degree, entries)

                    m.extend_by_zero(t - 1)
                    d.extend_by_zero(t - 1)

                    entries = [g]
                    cur_degree = t

            m.add_generators(cur_degree, len(entries), None)
            d.add_generators_from_rows(cur_degree, entries)

            m.extend_by_zero(max_n + s + 1)
            d.extend_by_zero(max_n + s)

    return (s, cc)


def main():
    data_dir = os.path.join(os.path.dirname(__file__), "bruner_data")
    max_n = query.with_default("Max n", "20", int)

    # Read in Bruner's resolution
    max_s, cc = read_bruner_resolution(data_dir, max_n)
    max = sseq.Bidegree.n_s(max_n, max_s)

    save_dir = query.optional("Save directory", str)

    resolution = Resolution.construct("S_2@milnor", save_dir)

    resolution.compute_through_stem(max)

    # Create a ResolutionHomomorphism object
    # NOTE: Bidegree::zero() is not bound (API_PROPOSAL §6.1 lists only s_t/n_s/x_y);
    # Bidegree.n_s(0, 0) is the faithful equivalent.
    hom = ResolutionHomomorphism("", cc, resolution, sseq.Bidegree.n_s(0, 0))

    # We have to explicitly tell it what to do at (0, 0)
    hom.extend_step(sseq.Bidegree.n_s(0, 0), Matrix.from_vec(TWO, [[1]]))
    hom.extend_all()

    # Now print the results
    print("sseq_basis | bruner_basis")
    for b in hom.target().iter_stem():
        matrix = hom.get_map(b.s).hom_k(b.t)

        for i, row in enumerate(matrix):
            g = sseq.BidegreeGenerator(b, i)
            print(f"x_{g:#} = {list(row)}")


if __name__ == "__main__":
    main()
